src/server/services/wmts/qgswmtsutils.cpp

```diff
@@ -0,0 +1,515 @@
+    }
+    else
+    {
+      throw QgsRequestNotWellFormedException( QStringLiteral( "TileCol is mandatory" ) );
+    }
+    if ( tm.col <= tc )
+    {
+      throw QgsRequestNotWellFormedException( QStringLiteral( "TileCol is unknown" ) );
+    }
+
+    int tileWidth = 256;
+    int tileHeight = 256;
+    double res = tm.resolution;
+    double minx = tm.left + tc * ( tileWidth * res );
+    double miny = tm.top - ( tr + 1 ) * ( tileHeight * res );
+    double maxx = tm.left + ( tc + 1 ) * ( tileWidth * res );
+    double maxy = tm.top - tr * ( tileHeight * res );
+    QString bbox;
+    if ( tms.ref == "EPSG:4326" )
+    {
+      bbox = qgsDoubleToString( miny, 6 ) + ',' +
+             qgsDoubleToString( minx, 6 ) + ',' +
+             qgsDoubleToString( maxy, 6 ) + ',' +
+             qgsDoubleToString( maxx, 6 );
+    }
+    else
+    {
+      bbox = qgsDoubleToString( minx, 6 ) + ',' +
+             qgsDoubleToString( miny, 6 ) + ',' +
+             qgsDoubleToString( maxx, 6 ) + ',' +
+             qgsDoubleToString( maxy, 6 );
+    }
+
+    QUrlQuery query;
+    if ( !params.value( QStringLiteral( "MAP" ) ).isEmpty() )
+    {
+      query.addQueryItem( QStringLiteral( "map" ), params.value( QStringLiteral( "MAP" ) ) );
+    }
+    query.addQueryItem( QStringLiteral( "service" ), QStringLiteral( "WMS" ) );
+    query.addQueryItem( QStringLiteral( "version" ), QStringLiteral( "1.3.0" ) );
+    query.addQueryItem( QStringLiteral( "request" ), request );
+    query.addQueryItem( QStringLiteral( "layers" ), layer );
+    query.addQueryItem( QStringLiteral( "styles" ), QString() );
+    query.addQueryItem( QStringLiteral( "crs" ), tms.ref );
+    query.addQueryItem( QStringLiteral( "bbox" ), bbox );
+    query.addQueryItem( QStringLiteral( "width" ), QStringLiteral( "256" ) );
+    query.addQueryItem( QStringLiteral( "height" ), QStringLiteral( "256" ) );
+    query.addQueryItem( QStringLiteral( "format" ), format );
+    if ( format.startsWith( QStringLiteral( "image/png" ) ) )
+    {
+      query.addQueryItem( QStringLiteral( "transparent" ), QStringLiteral( "true" ) );
+    }
+    query.addQueryItem( QStringLiteral( "dpi" ), QStringLiteral( "96" ) );
+
+    return query;
+  }
+
+  namespace
+  {
+
+    QMap< QgsUnitTypes::DistanceUnit, double> populateInchesPerUnit()
+    {
+      QMap< QgsUnitTypes::DistanceUnit, double>  m;
+      m[ QgsUnitTypes::DistanceMeters ] = 39.37;
+      m[ QgsUnitTypes::DistanceFeet ] = 12.0;
+      m[ QgsUnitTypes::DistanceYards ] = 36.0;
+      m[ QgsUnitTypes::DistanceMiles ] = 63360.0;
+      m[ QgsUnitTypes::DistanceDegrees ] = 4374754.0;
+      m[ QgsUnitTypes::DistanceKilometers ] = m[ QgsUnitTypes::DistanceMeters ] * 1000.0;
+      m[ QgsUnitTypes::DistanceNauticalMiles ] = m[ QgsUnitTypes::DistanceMeters ] * 1852.0;
+      m[ QgsUnitTypes::DistanceCentimeters ] = m[ QgsUnitTypes::DistanceMeters ] / 100.0;
+      m[ QgsUnitTypes::DistanceMillimeters ] = m[ QgsUnitTypes::DistanceMeters ] / 1000.0;
```

Comment:
I'm wondering if these constants should be defined in core instead of server? @nyalldawson what do you think?